ClassNotFoundException on path DexPathList

Seen from builds 1061, 1063 and 1065 in the Google Play Developer console. It has happened three times each on Android 4.2, 4.4 and 5.1 (so not very frequently). There is a recent user message that says "legal help page" but I don't know what that means.

````
java.lang.RuntimeException: Unable to instantiate application org.odk.collect.android.application.Collect: java.lang.ClassNotFoundException: Didn't find class "org.odk.collect.android.application.Collect" on path: DexPathList[[zip file "/system/framework/com.google.android.maps.jar"],nativeLibraryDirectories=[/vendor/lib, /system/lib]]
	at android.app.LoadedApk.makeApplication(LoadedApk.java:628)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4964)
	at android.app.ActivityThread.access$1600(ActivityThread.java:177)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1491)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:135)
	at android.app.ActivityThread.main(ActivityThread.java:5910)
	at java.lang.reflect.Method.invoke(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:372)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1405)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1200)
Caused by: java.lang.ClassNotFoundException: Didn't find class "org.odk.collect.android.application.Collect" on path: DexPathList[[zip file "/system/framework/com.google.android.maps.jar"],nativeLibraryDirectories=[/vendor/lib, /system/lib]]
	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
	at android.app.Instrumentation.newApplication(Instrumentation.java:987)
	at android.app.LoadedApk.makeApplication(LoadedApk.java:623)
	... 10 more
	Suppressed: java.lang.ClassNotFoundException: org.odk.collect.android.application.Collect
		at java.lang.Class.classForName(Native Method)
		at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
		at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
		at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
		... 13 more
Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack available
````

[This StackOverflow post](http://stackoverflow.com/questions/29816745/classnotfoundexception-didnt-find-class-on-path-dexpathlist) has some ideas to explore.

This is @shobhitagarwal1612's stack trace reported on Slack:

```
 02-04 01:51:04.067 9599-9599/org.odk.collect.android E/AndroidRuntime: FATAL EXCEPTION: main
                                                                        Process: org.odk.collect.android, PID: 9599
      java.lang.RuntimeException: Unable to get provider org.odk.collect.android.provider.InstanceProvider: java.lang.ClassNotFoundException: Didn't find class "org.odk.collect.android.provider.InstanceProvider" on path: DexPathList[[zip file "/data/app/org.odk.collect.android-1.apk"],nativeLibraryDirectories=[/data/app-lib/org.odk.collect.android, /vendor/lib, /system/lib]]
          at android.app.ActivityThread.installProvider(ActivityThread.java:5011)
          at android.app.ActivityThread.installContentProviders(ActivityThread.java:4582)
          at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4522)
          at android.app.ActivityThread.access$1500(ActivityThread.java:151)
          at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1381)
          at android.os.Handler.dispatchMessage(Handler.java:110)
          at android.os.Looper.loop(Looper.java:193)
          at android.app.ActivityThread.main(ActivityThread.java:5292)
          at java.lang.reflect.Method.invokeNative(Native Method)
          at java.lang.reflect.Method.invoke(Method.java:515)
          at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824)
          at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640)
          at dalvik.system.NativeStart.main(Native Method)
       Caused by: java.lang.ClassNotFoundException: Didn't find class "org.odk.collect.android.provider.InstanceProvider" on path: DexPathList[[zip file "/data/app/org.odk.collect.android-1.apk"],nativeLibraryDirectories=[/data/app-lib/org.odk.collect.android, /vendor/lib, /system/lib]]
          at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
          at java.lang.ClassLoader.loadClass(ClassLoader.java:497)
          at java.lang.ClassLoader.loadClass(ClassLoader.java:457)
          at android.app.ActivityThread.installProvider(ActivityThread.java:4996)
          at android.app.ActivityThread.installContentProviders(ActivityThread.java:4582) 
          at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4522) 
          at android.app.ActivityThread.access$1500(ActivityThread.java:151) 
          at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1381) 
          at android.os.Handler.dispatchMessage(Handler.java:110) 
          at android.os.Looper.loop(Looper.java:193) 
          at android.app.ActivityThread.main(ActivityThread.java:5292) 
          at java.lang.reflect.Method.invokeNative(Native Method) 
          at java.lang.reflect.Method.invoke(Method.java:515) 
          at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:824) 
          at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:640) 
          at dalvik.system.NativeStart.main(Native Method) 
 ```

--- a/collect_app/src/main/java/org/odk/collect/android/application/Collect.java
+++ b/collect_app/src/main/java/org/odk/collect/android/application/Collect.java
@@ -22,6 +22,7 @@ import android.net.ConnectivityManager;
 import android.net.NetworkInfo;
 import android.os.Environment;
 import android.preference.PreferenceManager;
+import android.support.multidex.MultiDex;
 
 import com.google.android.gms.analytics.GoogleAnalytics;
 import com.google.android.gms.analytics.Tracker;
@@ -170,6 +171,13 @@ public class Collect extends Application {
                 }
             }
         }
+    }
+
+
+    @Override
+    protected void attachBaseContext(Context base) {
+        super.attachBaseContext(base);
+        MultiDex.install(this);
     }
 
     /**

--- a/collect_app/src/main/java/org/odk/collect/android/application/Collect.java
+++ b/collect_app/src/main/java/org/odk/collect/android/application/Collect.java
@@ -52,10 +52,6 @@ import java.io.File;
  */
 public class Collect extends Application {
 
-    static {
-        PRNGFixes.apply();
-    }
-
     // Storage paths
     public static final String ODK_ROOT = Environment.getExternalStorageDirectory()
             + File.separator + "odk";
@@ -67,9 +63,13 @@ public class Collect extends Application {
     public static final String TMPDRAWFILE_PATH = CACHE_PATH + File.separator + "tmpDraw.jpg";
     public static final String TMPXML_PATH = CACHE_PATH + File.separator + "tmp.xml";
     public static final String LOG_PATH = ODK_ROOT + File.separator + "log";
-
     public static final String DEFAULT_FONTSIZE = "21";
     public static final String OFFLINE_LAYERS = ODK_ROOT + File.separator + "layers";
+    private static Collect singleton = null;
+
+    static {
+        PRNGFixes.apply();
+    }
 
     // share all session cookies across all sessions...
     private CookieStore cookieStore = new BasicCookieStore();
@@ -78,33 +78,10 @@ public class Collect extends Application {
     private ActivityLogger mActivityLogger;
     private FormController mFormController = null;
     private ExternalDataManager externalDataManager;
-
-    private static Collect singleton = null;
-
     private Tracker mTracker;
 
     public static Collect getInstance() {
         return singleton;
-    }
-
-    public ActivityLogger getActivityLogger() {
-        return mActivityLogger;
-    }
-
-    public FormController getFormController() {
-        return mFormController;
-    }
-
-    public void setFormController(FormController controller) {
-        mFormController = controller;
-    }
-
-    public ExternalDataManager getExternalDataManager() {
-        return externalDataManager;
-    }
-
-    public void setExternalDataManager(ExternalDataManager externalDataManager) {
-        this.externalDataManager = externalDataManager;
     }
 
     public static int getQuestionFontsize() {
@@ -114,27 +91,6 @@ public class Collect extends Application {
                 Collect.DEFAULT_FONTSIZE);
         int questionFontsize = Integer.valueOf(question_font);
         return questionFontsize;
-    }
-
-    public String getVersionedAppName() {
-        String versionName = "";
-        try {
-            versionName = getPackageManager()
-                    .getPackageInfo(getPackageName(), 0)
-                    .versionName;
-            versionName = " " + versionName.replaceFirst("-", "\n");
-        } catch (NameNotFoundException e) {
-            // TODO Auto-generated catch block
-            e.printStackTrace();
-        }
-        return getString(R.string.app_name) + versionName;
-    }
-
-    public boolean isNetworkAvailable() {
-        ConnectivityManager manager = (ConnectivityManager) getInstance()
-                .getSystemService(Context.CONNECTIVITY_SERVICE);
-        NetworkInfo currentNetworkInfo = manager.getActiveNetworkInfo();
-        return currentNetworkInfo != null && currentNetworkInfo.isConnected();
     }
 
     /**
@@ -173,13 +129,6 @@ public class Collect extends Application {
         }
     }
 
-
-    @Override
-    protected void attachBaseContext(Context base) {
-        super.attachBaseContext(base);
-        MultiDex.install(this);
-    }
-
     /**
      * Predicate that tests whether a directory path might refer to an
      * ODK Tables instance data directory (e.g., for media attachments).
@@ -199,6 +148,57 @@ public class Collect extends Application {
             }
         }
         return false;
+    }
+
+    public ActivityLogger getActivityLogger() {
+        return mActivityLogger;
+    }
+
+    public FormController getFormController() {
+        return mFormController;
+    }
+
+    public void setFormController(FormController controller) {
+        mFormController = controller;
+    }
+
+    public ExternalDataManager getExternalDataManager() {
+        return externalDataManager;
+    }
+
+    public void setExternalDataManager(ExternalDataManager externalDataManager) {
+        this.externalDataManager = externalDataManager;
+    }
+
+    public String getVersionedAppName() {
+        String versionName = "";
+        try {
+            versionName = getPackageManager()
+                    .getPackageInfo(getPackageName(), 0)
+                    .versionName;
+            versionName = " " + versionName.replaceFirst("-", "\n");
+        } catch (NameNotFoundException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+        return getString(R.string.app_name) + versionName;
+    }
+
+    public boolean isNetworkAvailable() {
+        ConnectivityManager manager = (ConnectivityManager) getInstance()
+                .getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo currentNetworkInfo = manager.getActiveNetworkInfo();
+        return currentNetworkInfo != null && currentNetworkInfo.isConnected();
+    }
+
+    /*
+        Adds support for multidex support library. For more info check out the link below,
+        https://developer.android.com/studio/build/multidex.html
+    */
+    @Override
+    protected void attachBaseContext(Context base) {
+        super.attachBaseContext(base);
+        MultiDex.install(this);
     }
 
     /**
@@ -257,6 +257,7 @@ public class Collect extends Application {
 
     /**
      * Gets the default {@link Tracker} for this {@link Application}.
+     *
      * @return tracker
      */
     synchronized public Tracker getDefaultTracker() {

